Tests use libvirt instead of qemu

dperpeet opened the pull request:
still an issue: after vm-install of the selinux-package, the virtual machine won't boot (selinux-error)

dperpeet commented:
fixes #2628

mvollmer commented:
Does running multiple machines from the same image at the same time work?

dperpeet commented:
> Does running multiple machines from the same image at the same time work?

no, [libvirt assumes the name to be unique](https://libvirt.org/formatdomain.html#elementsMetadata)
I added a patch to modifiy the name with a random extension and catch this specific error if it occurs (retry if necessary)

stefwalter commented:
Confirming that SELinux failure goes away if ./testsuite-prepare skips cockpit-selinux-policy. In addition when installing cockpit-selinux-policy on its own, in a later ./vm-run instance, we see:

```
Running transaction
  Installing  : cockpit-selinux-policy-wip-1.fc22.noarch                    1/1 
[  151.203290] SELinux:  Permission audit_read in class capability2 not defined in policy.
[  151.203965] SELinux:  Class binder not defined in policy.
[  151.204323] SELinux: the above unknown classes and permissions will be allowed
[  151.666000] audit_printk_skb: 3 callbacks suppressed
[  151.666386] audit: type=1403 audit(1442318299.337:489): policy loaded auid=0 ses=1
[  151.667244] audit: type=1300 audit(1442318299.337:489): arch=c000003e syscall=1 success=yes exit=3789379 a0=4 a1=7fe7c68b8000 a2=39d243 a3=85c items=0 ppid=1393 pid=1394 auid=0 uid=0 gid=0 euid=0 suid=0 fsuid=0 egid=0 sgid=0 fsgid=0 tty=ttyS0 ses=1 comm="load_policy" exe="/usr/sbin/load_policy" subj=unconfined_u:unconfined_r:load_policy_t:s0-s0:c0.c1023 key=(null)
[  151.669350] audit: type=1327 audit(1442318299.337:489): proctitle="/sbin/load_policy"
```

However when installing cockpit-selinux-policy during a later manual ./vm-run ... everything seems to work fine.

stefwalter commented:
What about:

```
-                                    "-o", "backing_file=%s" % self._image_image,
+                                    "-o", "backing_file=%s" % self._image_image, "-o", "backing_fmt=qcow2",
```

stefwalter commented:
So for anyone playing along. The problem here is that the libvirt domain (the VM) was killed before it had a chance to write its data to disk. Adding the following lines to ```debug-install_packages``` made the symptoms go away:

```
sync
sync
sync
sync
```

stefwalter commented:
More evidence. Even when i booted without selinux enabled (on such a screwed up vm instance) ... the boot would fail here:

```
[    6.600512] bridge: automatic filtering via arp/ip/ip6tables has been deprecated. Update your scripts to load br_netfilter if you need this.
[   56.516076] audit_printk_skb: 207 callbacks suppressed
[   56.516510] audit: type=1130 audit(1442316259.686:69): pid=1 uid=0 auid=4294967295 ses=4294967295 subj=kernel msg='unit=dbus comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
[   56.519669] audit: type=1130 audit(1442316259.689:70): pid=1 uid=0 auid=4294967295 ses=4294967295 subj=kernel msg='unit=systemd-logind comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=failed'
         Starting D-Bus System Message Bus...
         Starting GSSAPI Proxy Daemon...
[  OK  ] Reached target Timers.
[FAILED] Failed to start Login Service.
[   56.524333] audit: type=1131 audit(1442316259.694:71): pid=1 uid=0 auid=4294967295 ses=4294967295 subj=kernel msg='unit=dbus comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=failed'
 
[   56.530431] audit: type=1130 audit(1442316259.700:72): pid=1 uid=0 auid=4294967295 ses=4294967295 subj=kernel msg='unit=gssproxy comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
[  OK  ] Started GSSAPI Proxy Daemon.
[**    ] A start job is running for firewall...all daemon (1min 30s / 1min 30s)[   97.404266] audit: type=1130 audit(1442316300.574:73): pid=1 uid=0 auid=4294967295 ses=4294967295 subj=kernel msg='unit=firewalld comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=failed'
[FAILED] Failed to start firewalld - dynamic firewall daemon.
See "systemctl status firewalld.service" for details.
```

And firewall-cmd is one of the last things that ```debug-install_packages``` touches.

dperpeet commented:
using `virDomainShutdown` seems to do the trick!

dperpeet commented:
most tests should work now, there are still issues with adding a network interface

stefwalter commented:
@dperpeet I think you should just use the API equivalent of this for now. Don't worry about reworking the add network interface stuff, just run it the way it did before:

```
virsh qemu-monitor-command --hmp <domain> '<command> [...]'
```

dperpeet commented:
forgot to update text here: network tests work, it's down to adding storage with speed limitation (nbd: socket and trickle)

mvollmer commented:
> it's down to adding storage with speed limitation (nbd: socket and trickle)

It would be best to find another way to test jobs, and not try to create slow disks.  I would go as far as mocking the storaged D-Bus API, I think.

mvollmer commented:
That being said, it would also be great to test our graphs with real numbers.  E.e.g, add a disk with specific speed and then try to get that number in the plots.  Same for network, etc.

dperpeet commented:
> That being said, it would also be great to test our graphs with real numbers. E.e.g, add a disk with specific speed and then try to get that number in the plots. Same for network, etc.

The libvirt API has this functionality built in. I exposed it [in testvm.py](https://github.com/cockpit-project/cockpit/pull/2759/files#diff-ad31112db9347f7ac1d8207975a36a5dR916)

dperpeet commented:
added mutex protection to vm status update event handling, changed waiting for boot/reboot to use signals instead of polling for new events

@mvollmer right now I store the current domain state and update the information as events come in. Would it be enough to sort the events into an event queue for a domain and let the main thread consume those as it waits for a certain state to occur? I believe this could be enough since we only have one consumer and usually only wait for something specific to happen after we induced a change (i.e. the previous state doesn't matter in that case).

dperpeet commented:
current problems:
 * docker doesn't work in fedora 22 image
 * check-realms fails:

```
+ sleep 1
+ for x in 1 2 3 4 5 6 7 8 9 10
+ nslookup -type=SRV _ldap._tcp.cockpit.lan
Server:		10.111.112.100
Address:	10.111.112.100#53

** server can't find _ldap._tcp.cockpit.lan: NXDOMAIN

+ sleep 1
+ realm join -U admin cockpit.lan
+ echo foobarfoo
realm: No such realm found
ERROR

======================================================================
ERROR: testNegotiate (__main__.TestKerberos)
```

dperpeet commented:
the above issues have all been addressed

stefwalter commented:
I get this issue when running on files.cockpit-project.org:

```
Running setup script /home/hubbot/hubbot/fc843ce09b20b1df833b6254ab259e61a5a395ec_f22_x86-64/test/guest/openshift-fedora-22.setup
libvirt: Storage Driver error : Cannot access storage file '/home/hubbot/hubbot/fc843ce09b20b1df833b6254ab259e61a5a395ec_f22_x86-64/test/run/openshift-fedora-22-x86_64-1.qcow2' (as uid:107, gid:107): Permission denied
Traceback (most recent call last):
  File "./vm-create", line 233, in <module>
    builder.build(args.args)
  File "./vm-create", line 220, in build
    self.run_setup_script(script, args)
  File "./vm-create", line 160, in run_setup_script
    self.machine.start(maintain=True)
  File "/home/hubbot/hubbot/fc843ce09b20b1df833b6254ab259e61a5a395ec_f22_x86-64/test/testvm.py", line 844, in start
    self._domain = self._start_qemu(maintain=maintain, macaddr=macaddr)
  File "/home/hubbot/hubbot/fc843ce09b20b1df833b6254ab259e61a5a395ec_f22_x86-64/test/testvm.py", line 793, in _start_qemu
    dom = self.virt_connection.createXML(test_domain_desc, libvirt.VIR_DOMAIN_START_AUTODESTROY | libvirt.VIR_DOMAIN_START_PAUSED)
  File "/usr/lib64/python2.7/site-packages/libvirt.py", line 3427, in createXML
    if ret is None:raise libvirtError('virDomainCreateXML() failed', conn=self)
libvirt.libvirtError: Cannot access storage file '/home/hubbot/hubbot/fc843ce09b20b1df833b6254ab259e61a5a395ec_f22_x86-64/test/run/openshift-fedora-22-x86_64-1.qcow2' (as uid:107, gid:107): Permission denied
```

I simply ran ```./VERIFY``` in ```/home/hubbot/hubbot/fc843ce09b20b1df833b6254ab259e61a5a395ec_f22_x86-64``` as user ```hubbot```.

dperpeet commented:
@stefwalter your VERIFY failed, because the images aren't all properly created yet
hubbot killed the creation due to 60 minute timeout

stefwalter commented:
> @stefwalter your VERIFY failed, because the images aren't all properly created yet
hubbot killed the creation due to 60 minute timeout

I'm running manually, as noted above in that directory ... without a timeout.

stefwalter commented:
Exact reproduction instructions:

```
$ ssh files.cockpit-project.org
$ sudo -s
# su hubbot
$ cd ~hubbot/hubbot/fc843ce09b20b1df833b6254ab259e61a5a395ec_f22_x86-64
$ ./VERIFY
```

stefwalter commented:
Turning off selinux has no effect.

stefwalter commented on test/testvm.py:
```diff
@@ -475,117 +600,39 @@ def __init__(self, **args):
 
-    def unpack_base(self, modify_func=None):
-        assert not self._process
+        self.virt_connection = libvirt.open()
```
You need to pass "qemu:///session" here, to get predictable behavior here.

stefwalter commented:
I got this failure. Any ideas?

```
======================================================================
ERROR: testBasic (__main__.TestShutdownRestart)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./check-shutdown-restart", line 44, in testBasic
    m.wait_reboot()
  File "/home/hubbot/hubbot/c5b4046985ad76a818a7c75404c5f91c4fb8bd47_f22_x86-64/test/testvm.py", line 898, in wait_reboot
    raise Failure("system didn't reboot properly")
Failure: system didn't reboot properly

----------------------------------------------------------------------
Ran 1 test in 151.048s

```

stefwalter commented:
> I got this failure. Any ideas?

This is reliably reproducible. I can give you instructions on IRC.

dperpeet commented:
`VIR_DOMAIN_EVENT_ID_REBOOT` - the callback function for this event is called too late for some reason (only once we've finished waiting)

dperpeet commented:
libvirt: The event loop needs to be initialized before opening a connection to a session.
I couldn't find this in the [documentation](http://www.libvirt.org/html/libvirt-libvirt-event.html#virEventRegisterDefaultImpl), but it's stated in an old [e-mail](https://www.redhat.com/archives/libvirt-users/2014-April/msg00011.html). I will propose a patch to that documentation if this really was the issue.

stefwalter commented:
These SELinux failures on RHEL seem new:

```
OK
./check-system-info
Unexpected journal message 'Sent message type=method_call sender=n/a destination=org.freedesktop.DBus object=/org/freedesktop/DBus interface=org.freedesktop.DBus member=Hello cookie=1 reply_cookie=0 error=n/a'
Unexpected journal message 'Got message type=method_return sender=org.freedesktop.DBus destination=:1.31 object=n/a interface=n/a member=n/a cookie=1 reply_cookie=1 error=n/a'
Unexpected journal message 'Sent message type=method_call sender=n/a destination=org.freedesktop.DBus object=/org/freedesktop/DBus interface=org.freedesktop.DBus member=Hello cookie=1 reply_cookie=0 error=n/a'
Unexpected journal message 'Got message type=method_return sender=org.freedesktop.DBus destination=:1.42 object=n/a interface=n/a member=n/a cookie=1 reply_cookie=1 error=n/a'
Unexpected journal message 'Sent message type=method_call sender=n/a destination=org.freedesktop.DBus object=/org/freedesktop/DBus interface=org.freedesktop.DBus member=Hello cookie=1 reply_cookie=0 error=n/a'
Unexpected journal message 'Got message type=method_return sender=org.freedesktop.DBus destination=:1.50 object=n/a interface=n/a member=n/a cookie=1 reply_cookie=1 error=n/a'
Journal database copied to check-system-info-testTime-10.111.112.92-FAIL.journal
.Warning: Permanently added '10.111.112.92' (ECDSA) to the list of known hosts.
E
======================================================================
ERROR: testTime (__main__.TestSystemInfo)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hubbot/hubbot/89c1b2130242d8a282ea917a26ad6b8d3916db6b_r7_x86-64/test/testlib.py", line 435, in tearDown
    self.check_journal_messages()
  File "/home/hubbot/hubbot/89c1b2130242d8a282ea917a26ad6b8d3916db6b_r7_x86-64/test/testlib.py", line 564, in check_journal_messages
    raise Error("There were unexpected journal messages")
Error: There were unexpected journal messages

----------------------------------------------------------------------
Ran 2 tests in 106.952s

FAILED (errors=1)
./check-networking
Unexpected journal message 'Successfully loaded SELinux database in 24.800ms, size on heap is 92K.'
Unexpected journal message 'calling: info'
Unexpected journal message 'Successfully loaded SELinux database in 7.468ms, size on heap is 92K.'
Unexpected journal message 'calling: info'
Unexpected journal message 'Successfully loaded SELinux database in 45.404ms, size on heap is 92K.'
Unexpected journal message 'calling: info'
Journal database copied to check-networking-testBasic-10.111.112.102-FAIL.journal
Warning: Permanently added '10.111.112.102' (ECDSA) to the list of known hosts.
E
======================================================================
ERROR: testBasic (__main__.TestNetworking)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hubbot/hubbot/89c1b2130242d8a282ea917a26ad6b8d3916db6b_r7_x86-64/test/testlib.py", line 435, in tearDown
    self.check_journal_messages()
  File "/home/hubbot/hubbot/89c1b2130242d8a282ea917a26ad6b8d3916db6b_r7_x86-64/test/testlib.py", line 564, in check_journal_messages
    raise Error("There were unexpected journal messages")
Error: There were unexpected journal messages
```

http://files.cockpit-project.org/hubbot/89c1b2130242d8a282ea917a26ad6b8d3916db6b_r7_x86-64/hubbot.html

stefwalter commented:
I have a fix for testBond.

stefwalter commented:
Could we split out all the image changes into a separate pull request? That way they'll be easier to review, and track what has actually changed.

dperpeet commented:
> Could we split out all the image changes into a separate pull request? That way they'll be easier to review, and track what has actually changed.

All the stuff is highly dependent. We may be able to keep all images except ipa... what about openshift?

stefwalter commented:
testBond fails because of the image changes.

stefwalter commented:
Two additional commits pushed to stefwalter/test_libvirt

stefwalter commented:
./vm-run fails:

```
[hubbot@files test]$ ./vm-run 
console, to quit use Ctrl+], Ctrl+5 (depending on locale)
error: failed to get domain '2'
error: Domain not found: no domain with matching name '2'
```

stefwalter commented on test/check-networking:
```diff
@@ -19,15 +19,35 @@
 
     def get_iface(self, m, mac):
         def getit():
             return m.execute("for f in /sys/class/net/*; do a=$(cat $f/address); if test \"$a\" == %s; then echo $(basename $f); fi; done" % mac)
```
I pushed a change to use grep instead of this. Should I put it on a separate branch?

dperpeet commented:
> [hubbot@files test]$ ./vm-run console, to quit use Ctrl+], Ctrl+5 (depending on locale) error: failed to get domain '2' error: Domain not found: no domain with matching name '2'

that error occurs if virsh can't find the machine
did the vm crash?

stefwalter commented on test/check-networking:
```diff
@@ -133,17 +152,11 @@ class TestNetworking(MachineCase):
-            b.wait_in_text("#networking-interfaces tr:contains('%s')" % iface2, "Configuring IP")
+        mac = "52:54:00:9e:00:F5"
+        iface1 = self.get_iface(m, m.add_netiface(mac=mac));
+        iface2 = self.get_iface(m, m.add_netiface(secondary_network=True));
```
I've pushed a commit that removes the secondary_network stuff and just uses QEMU network monitor defined networks. It has less litter, doesn't need root permissions etc.

stefwalter commented on test/guest/cockpit.conf:
```diff
@@ -1,7 +1,7 @@
-            "fedora-atomic-22":  "4",
-            "fedora-rawhide":    "7",
-            "fedora-22-testing": "6"
+{ "tags": { "fedora-22":         "22",
```
The new images break testBond test. With the old images, that test works. Not sure why that is.

stefwalter commented on test/guest/ipa-fedora-22.setup:
```diff
@@ -10,18 +10,9 @@ iptables -F
-
-ln -sf ../selinux/config /etc/sysconfig/selinux
-echo 'SELINUX=permissive' > /etc/selinux/config
+dnf clean all
```
Why is this line after systemctl daemon-rexec in one setup, and before it in another?

stefwalter commented on test/guest/cockpit-fedora-22.setup:
```diff
@@ -44,24 +44,9 @@ dnf -y upgrade dnf
+# bug in selinux-policy-targeted-3.13.1-128.13.fc22
+# HACK https://bugzilla.redhat.com/show_bug.cgi?id=1224211
+# we need to restart systemctl and then reboot the system
+systemctl daemon-reexec
```
This seems like something we should just hard code into vm-create for now. It doesn't need to be in every setup script.

stefwalter commented on test/testvm.py:
```diff
@@ -21,18 +21,22 @@
 import errno
 import fcntl
 import guestfs
```
guestfs is no longer needed.

stefwalter commented on test/testvm.py:
```diff
@@ -475,117 +604,42 @@ def __init__(self, **args):
-            os.makedirs(self.run_dir, 0750)
+        # network names are currently hardcoded into network-cockpit.xml and network-cockpit_secondary.xml
+        self.network_name = self._read_network_name()
+        self.network_name_secondary = self._read_network_name_secondary()
```
Same as above about secondary networks. I have a commit that removes them at stefwalter/test_libvirt.

stefwalter commented on test/testvm.py:
```diff
@@ -664,19 +708,29 @@ def _choose_macaddr(self):
-    def _start_qemu(self, maintain=False, tty=False, monitor=None):
+        start_time = time.time()
+        # wait a bit before we give up
+        while (time.time() - start_time) < timeout_sec:
```
Why are we waiting 30 seconds to get a mac address for a flavor?

Alternate flavors are only used by a single test. Previously we would just fail. It's pretty inherent in our tests that you can't run the same test twice at the same time. Even in this pull request we have statically assigned mac addresses throughout many tests.

dperpeet replied:
One example is check-realms: All sub-tests need a clean ipa instance. We either give them individual mac addresses or allow a bit of time here - otherwise we set ourselves up for a race. When one test finishes, the virtual machines are killed and libvirt / the os are left to clean up the mess. This includes the lock file we create for the mac address. If the new test starts before the old one is cleaned up, we get a failure. I saw this happen on the test server during previous runs.

dperpeet replied:
We should comment this accordingly, I believe.

stefwalter commented on test/testvm.py:
```diff
@@ -684,178 +738,233 @@ def _start_qemu(self, maintain=False, tty=False, monitor=None):
+                else:
+                    raise
+            dom_created = True
+
```
Could we set self._domain here, rather than relying on every caller to do it?

stefwalter commented on test/testvm.py:
```diff
@@ -684,178 +738,233 @@ def _start_qemu(self, maintain=False, tty=False, monitor=None):
+            if self.flavor == "ipa":
+                self.macaddr = "52:54:00:9e:00:F0"
+            elif self.flavor == "openshift":
+                self.macaddr = "52:54:00:9e:00:F1"
```
Shouldn't these come from the network definition?

stefwalter commented on test/testvm.py:
```diff
@@ -684,178 +738,233 @@ def _start_qemu(self, maintain=False, tty=False, monitor=None):
+            if self.flavor == "ipa":
+                self.macaddr = "52:54:00:9e:00:F0"
+            elif self.flavor == "openshift":
+                self.macaddr = "52:54:00:9e:00:F1"
```
Shouldn't these come from the network definition?

stefwalter commented on test/testvm.py:
```diff
@@ -684,178 +738,233 @@ def _start_qemu(self, maintain=False, tty=False, monitor=None):
+    def wait_reboot(self):
+        if not self.event_handler.wait_for_reboot(self._domain):
+            raise Failure("system didn't notify us about a reboot")
+        # we may have to check for a new dhcp lease, but the old one can be active for a bit
```
This seems like code duplicated with wait_ssh() Could we share code here?

stefwalter commented on test/testvm.py:
```diff
@@ -935,25 +1058,31 @@ def rem_disk(self, index):
-        return mac
+    def add_netiface(self, mac=None, secondary_network=False):
+        desc_template = ""
+        with open("./files/test_domain_network.xml", "r") as desc_file:
```
I pushed some code to stefwalter/test_libvirt to do this without secondary networks in a more flexible way.

stefwalter commented:
> that error occurs if virsh can't find the machine did the vm crash?

You can test this yourself on files.cockpit-project.org in the hubbot user's ~/cockpit2 directory.

dperpeet commented:
all above issues have been addressed, ./vm-run crash hasn't been reproduced

dperpeet commented:
all above issues have been addressed

stefwalter commented:
Looking good.

But why are there still ```.1.setup``` scripts? I thought that commit bcf02a533b58928e343bc8f7d2a6ab83745e1b76 would have removed the need for them.

dperpeet commented:
> But why are there still .1.setup scripts? I thought that commit bcf02a5 would have removed the need for them.

`systemctl start docker` will fail after updating the system. Since our setup mechanics don't support waiting for a reboot (if the run command disconnects it's considered a failure), we still need the separate setup steps.

dperpeet commented:
failure for rhel 7, will need to check if I can reproduce this:

```
vm-install: Can't resolve IP of 52:54:00:5b:09:3a
Traceback (most recent call last):
  File "/home/hubbot/prod/hubbot/hubbot", line 584, in verify
    cmd("timeout", "120m", "./VERIFY")
  File "/home/hubbot/prod/hubbot/hubbot", line 462, in cmd
    subprocess.check_call(list(args), stdout=sys.stdout, stderr=sys.stderr)
  File "/usr/lib64/python2.7/subprocess.py", line 542, in check_call
    raise CalledProcessError(retcode, cmd)
CalledProcessError: Command '['timeout', '120m', './VERIFY']' returned non-zero exit status 1
```

dperpeet commented:
check-network fails on rhel-7, but I think this can be handled in a follow-up

dperpeet commented:
@mvollmer there is a more general failure, somehow tied to rhel 7:
```
  File "/home/hubbot/hubbot/778c66b98c70874ac05c9ec508d14620009fbe9b_r7_x86-64/test/testvm.py", line 842, in _wait_for_ip
    raise Failure("Can't resolve IP of %s" % mac)
Failure: Can't resolve IP of 52:54:00:e3:b2:7a
```
follow-up or solve now?

dperpeet commented:
the dhcp server may be running out of ip addresses to lease... we could expand the range

dperpeet commented:
Due to the significant number of active machines running tests, dhcp leases weren't expiring quickly enough. A larger address pool should take care of this.

stefwalter commented:
Unfortunately this test failure on f22 doesn't look good:

```
vm-install: Unable to reach machine 10.111.111.10 via ssh.
Traceback (most recent call last):
  File "/home/hubbot/prod/hubbot/hubbot", line 584, in verify
    cmd("timeout", "120m", "./VERIFY")
  File "/home/hubbot/prod/hubbot/hubbot", line 462, in cmd
    subprocess.check_call(list(args), stdout=sys.stdout, stderr=sys.stderr)
  File "/usr/lib64/python2.7/subprocess.py", line 542, in check_call
    raise CalledProcessError(retcode, cmd)
CalledProcessError: Command '['timeout', '120m', './VERIFY']' returned non-zero exit status 1
```

dperpeet commented:
The only remaining failure in rhel 7 check seems genuine:

```
./check-networking
Wrote check-networking-testBasic-FAIL.png
Journal database copied to check-networking-testBasic-10.111.112.102-FAIL.journal
EWarning: Permanently added '10.111.112.102' (ECDSA) to the list of known hosts.

======================================================================
ERROR: testBasic (__main__.TestNetworking)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./check-networking", line 133, in testBasic
    b.wait_in_text("tr:contains('IPv4')", "Automatic (DHCP)")
  File "/home/hubbot/hubbot/c79d28d32168eb42674249e6ef6b94c8d8cb7c17_r7_x86-64/test/testlib.py", line 238, in wait_in_text
    return self.wait_js_func('ph_in_text', selector, text)
  File "/home/hubbot/hubbot/c79d28d32168eb42674249e6ef6b94c8d8cb7c17_r7_x86-64/test/testlib.py", line 208, in wait_js_func
    return self.phantom.wait("%s(%s)" % (func, ','.join(map(jsquote, args))), timeout=self.phantom_wait_timeout)
  File "/home/hubbot/hubbot/c79d28d32168eb42674249e6ef6b94c8d8cb7c17_r7_x86-64/test/testlib.py", line 658, in wait
    return self.run({'cmd': 'wait', 'cond': cond, 'timeout': timeout*1000})
  File "/home/hubbot/hubbot/c79d28d32168eb42674249e6ef6b94c8d8cb7c17_r7_x86-64/test/testlib.py", line 621, in run
    raise Error("timeout")
Error: timeout

----------------------------------------------------------------------
Ran 1 test in 101.348s
```
